File: arc-compiler/src/lexer.rs
/// Arc lexer - ported from src/lexer.js

#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    // Literals
    STRING,
    NUMBER,
    BOOL,
    NONE,
    // Identifiers
    IDENT,
    AT_IDENT,
    // Structure keywords
    PAGE,
    WIDGET,
    DESIGN,
    IMPORT,
    EXPORT,
    FROM,
    // Logic keywords
    CONST,
    LET,
    FN,
    CLASS,
    RETURN,
    IF,
    ELSE,
    UNLESS,
    MATCH,
    FOR,
    IN,
    WHILE,
    UNTIL,
    LOOP,
    BREAK,
    CONTINUE,
    AWAIT,
    ASYNC,
    GEN,
    YIELD,
    TRY,
    CATCH,
    THROW,
    NEW,
    STATIC,
    GET,
    IS,
    RAW,
    // Result types
    OK,
    ERR,
    // Backend keywords
    MODEL,
    JOB,
    // Symbols
    LPAREN,
    RPAREN,
    LBRACE,
    RBRACE,
    LBRACKET,
    RBRACKET,
    DOT,
    DOTDOT,
    DOTDOTEQ,
    COMMA,
    COLON,
    SEMI,
    ARROW,
    THIN_ARROW,
    PIPE,
    QUESTION,
    QQDOT,
    QQ,
    HASH,
    SPREAD,
    // Operators
    PLUS,
    MINUS,
    STAR,
    SLASH,
    PERCENT,
    STARSTAR,
    EQ,
    EQEQ,
    BANGEQ,
    LT,
    GT,
    LTEQ,
    GTEQ,
    BANG,
    AMP,
    AMPAMP,
    BAR,
    BARBAR,
    PLUS_EQ,
    MINUS_EQ,
    STAR_EQ,
    SLASH_EQ,
    // Template-specific
    INDENT,
    DEDENT,
    NEWLINE,
    // Interpolation
    INTERP_START,
    INTERP_END,
    EOF,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenValue {
    None,
    Str(String),
    Num(f64),
    Bool(bool),
}

#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokenType,
    pub value: TokenValue,
    pub line: usize,
    pub col: usize,
}

impl Token {
    pub fn new(kind: TokenType, value: TokenValue, line: usize, col: usize) -> Self {
        Token { kind, value, line, col }
    }

    pub fn str_val(&self) -> &str {
        match &self.value {
            TokenValue::Str(s) => s.as_str(),
            _ => "",
        }
    }

    pub fn num_val(&self) -> f64 {
        match &self.value {
            TokenValue::Num(n) => *n,
            _ => 0.0,
        }
    }
}

fn keyword(s: &str) -> Option<TokenType> {
    match s {
        "page" => Some(TokenType::PAGE),
        "widget" => Some(TokenType::WIDGET),
        "design" => Some(TokenType::DESIGN),
        "import" => Some(TokenType::IMPORT),
        "export" => Some(TokenType::EXPORT),
        "from" => Some(TokenType::FROM),
        "const" => Some(TokenType::CONST),
        "let" => Some(TokenType::LET),
        "fn" => Some(TokenType::FN),
        "class" => Some(TokenType::CLASS),
        "return" => Some(TokenType::RETURN),
        "if" => Some(TokenType::IF),
        "else" => Some(TokenType::ELSE),
        "unless" => Some(TokenType::UNLESS),
        "match" => Some(TokenType::MATCH),
        "for" => Some(TokenType::FOR),
        "in" => Some(TokenType::IN),
        "while" => Some(TokenType::WHILE),
        "until" => Some(TokenType::UNTIL),
        "loop" => Some(TokenType::LOOP),
        "break" => Some(TokenType::BREAK),
        "continue" => Some(TokenType::CONTINUE),
        "await" => Some(TokenType::AWAIT),
        "async" => Some(TokenType::ASYNC),
        "gen" => Some(TokenType::GEN),
        "yield" => Some(TokenType::YIELD),
        "try" => Some(TokenType::TRY),
        "catch" => Some(TokenType::CATCH),
        "throw" => Some(TokenType::THROW),
        "new" => Some(TokenType::NEW),
        "static" => Some(TokenType::STATIC),
        "get" => Some(TokenType::GET),
        "is" => Some(TokenType::IS),
        "raw" => Some(TokenType::RAW),
        "true" => Some(TokenType::BOOL),
        "false" => Some(TokenType::BOOL),
        "none" => Some(TokenType::NONE),
        "Ok" => Some(TokenType::OK),
        "Err" => Some(TokenType::ERR),
        "model" => Some(TokenType::MODEL),
        "job" => Some(TokenType::JOB),
        _ => None,
    }
}

struct Lexer {
    source: Vec<char>,
    filename: String,
    pos: usize,
    line: usize,
    col: usize,
    tokens: Vec<Token>,
    indent_stack: Vec<usize>,
}

impl Lexer {
    fn new(source: &str, filename: &str) -> Self {
        Lexer {
            source: source.chars().collect(),
            filename: filename.to_string(),
            pos: 0,
            line: 1,
            col: 1,
            tokens: Vec::new(),
            indent_stack: vec![0],
        }
    }

    fn peek(&self, offset: usize) -> Option<char> {
        self.source.get(self.pos + offset).copied()
    }

    fn advance(&mut self) -> Option<char> {
        let ch = self.source.get(self.pos).copied()?;
        self.pos += 1;
        if ch == '\n' {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        Some(ch)
    }

    fn match_char(&mut self, expected: char) -> bool {
        if self.source.get(self.pos) == Some(&expected) {
            self.advance();
            true
        } else {
            false
        }
    }

    fn emit(&mut self, kind: TokenType, value: TokenValue, line: usize, col: usize) {
        self.tokens.push(Token::new(kind, value, line, col));
    }

    fn emit_at_current(&mut self, kind: TokenType, value: TokenValue) {
        let line = self.line;
        let col = self.col.saturating_sub(1);
        self.tokens.push(Token::new(kind, value, line, col));
    }

    fn skip_inline_whitespace(&mut self) {
        while self.pos < self.source.len() {
            match self.source[self.pos] {
                ' ' | '\t' => { self.advance(); }
                _ => break,
            }
        }
    }

    fn skip_line_comment(&mut self) {
        while self.pos < self.source.len() && self.source[self.pos] != '\n' {
            self.advance();
        }
    }

    fn measure_indent(&mut self) -> usize {
        let mut indent = 0;
        while self.pos < self.source.len() {
            match self.source[self.pos] {
                ' ' => { indent += 1; self.advance(); }
                '\t' => { indent += 2; self.advance(); }
                _ => break,
            }
        }
        indent
    }

    fn tokenize_string(&mut self, quote: char) {
        let line = self.line;
        let col = self.col.saturating_sub(1);
        let mut value = String::new();

        while self.pos < self.source.len() {
            let ch = self.source[self.pos];

            if ch == '\\' {
                self.advance();
                if let Some(esc) = self.advance() {
                    let escaped = match esc {
                        'n' => '\n',
                        't' => '\t',
                        'r' => '\r',
                        '\\' => '\\',
                        '"' => '"',
                        '\'' => '\'',
                        '{' => '{',
                        _ => esc,
                    };
                    value.push(escaped);
                }
                continue;
            }

            if quote == '"' && ch == '{' {
                // Start interpolation
                self.emit(TokenType::STRING, TokenValue::Str(value.clone()), line, col);
                value.clear();
                self.advance(); // consume {
                let interp_line = self.line;
                let interp_col = self.col.saturating_sub(1);
                self.emit(TokenType::INTERP_START, TokenValue::Str("{".to_string()), interp_line, interp_col);
                self.tokenize_interpolation();
                let end_line = self.line;
                let end_col = self.col.saturating_sub(1);
                self.emit(TokenType::INTERP_END, TokenValue::Str("}".to_string()), end_line, end_col);
                continue;
            }

            if ch == quote {
                self.advance();
                break;
            }

            if ch == '\n' && quote != '`' {
                // unterminated - break gracefully
                break;
            }

            value.push(ch);
            self.advance();
        }

        self.emit(TokenType::STRING, TokenValue::Str(value), line, col);
    }

    fn tokenize_interpolation(&mut self) {
        let mut depth = 0usize;
        while self.pos < self.source.len() {
            let ch = self.source[self.pos];
            if ch == '{' {
                depth += 1;
                self.tokenize_one();
            } else if ch == '}' {
                if depth == 0 {
                    self.advance();
                    return;
                }
                depth -= 1;
                self.tokenize_one();
            } else {
                self.tokenize_one();
            }
        }
    }

    fn tokenize_number(&mut self) {
        let start = self.pos - 1;
        while self.pos < self.source.len() {
            match self.source[self.pos] {
                '0'..='9' | '_' => { self.advance(); }
                _ => break,
            }
        }
        if self.pos < self.source.len() && self.source[self.pos] == '.' {
            if let Some(next) = self.source.get(self.pos + 1) {
                if next.is_ascii_digit() {
                    self.advance(); // consume .
                    while self.pos < self.source.len() {
                        match self.source[self.pos] {
                            '0'..='9' | '_' => { self.advance(); }
                            _ => break,
                        }
                    }
                }
            }
        }
        let raw: String = self.source[start..self.pos].iter().filter(|&&c| c != '_').collect();
        let num: f64 = raw.parse().unwrap_or(0.0);
        let line = self.line;
        let col = self.col.saturating_sub(self.pos - start);
        self.emit(TokenType::NUMBER, TokenValue::Num(num), line, col);
    }

    fn tokenize_ident(&mut self, first: char) {
        let start = self.pos - 1;
        while self.pos < self.source.len() {
            match self.source[self.pos] {
                'a'..='z' | 'A'..='Z' | '0'..='9' | '_' | '$' => { self.advance(); }
                _ => break,
            }
        }
        let value: String = self.source[start..self.pos].iter().collect();
        let line = self.line;
        let col = self.col.saturating_sub(self.pos - start);
        if let Some(kw) = keyword(&value) {
            let tv = match kw {
                TokenType::BOOL => {
                    if value == "true" { TokenValue::Bool(true) } else { TokenValue::Bool(false) }
                }
                _ => TokenValue::Str(value.clone()),
            };
            self.emit(kw, tv, line, col);
        } else {
            self.emit(TokenType::IDENT, TokenValue::Str(value), line, col);
        }
    }

    fn tokenize_at_sign(&mut self) {
        let start = self.pos - 1;
        while self.pos < self.source.len() {
            match self.source[self.pos] {
                'a'..='z' | 'A'..='Z' | '0'..='9' | '_' => { self.advance(); }
                _ => break,
            }
        }
        let value: String = self.source[start..self.pos].iter().collect();
        let line = self.line;
        let col = self.col.saturating_sub(self.pos - start);
        self.emit(TokenType::AT_IDENT, TokenValue::Str(value), line, col);
    }

    fn tokenize_one(&mut self) -> bool {
        self.skip_inline_whitespace();
        if self.pos >= self.source.len() {
            self.emit_at_current(TokenType::EOF, TokenValue::None);
            return false;
        }

        let ch = match self.advance() {
            Some(c) => c,
            None => return false,
        };
        let line = self.line;
        let col = self.col.saturating_sub(1);

        if ch == '\n' {
            self.emit(TokenType::NEWLINE, TokenValue::Str("\n".to_string()), line, col);
            return true;
        }

        if ch == '/' && self.peek(0) == Some('/') {
            self.skip_line_comment();
            return true;
        }

        if ch == '"' || ch == '\'' {
            self.tokenize_string(ch);
            return true;
        }

        if ch.is_ascii_digit() {
            self.tokenize_number();
            return true;
        }

        if ch.is_alphabetic() || ch == '_' || ch == '$' {
            self.tokenize_ident(ch);
            return true;
        }

        if ch == '@' {
            self.tokenize_at_sign();
            return true;
        }

        match ch {
            '.' => {
                if self.peek(0) == Some('.') && self.peek(1) == Some('.') {
                    self.advance(); self.advance();
                    self.emit(TokenType::SPREAD, TokenValue::Str("...".to_string()), line, col);
                } else if self.peek(0) == Some('.') && self.peek(1) == Some('=') {
                    self.advance(); self.advance();
                    self.emit(TokenType::DOTDOTEQ, TokenValue::Str("..=".to_string()), line, col);
                } else if self.peek(0) == Some('.') {
                    self.advance();
                    self.emit(TokenType::DOTDOT, TokenValue::Str("..".to_string()), line, col);
                } else {
                    self.emit(TokenType::DOT, TokenValue::Str(".".to_string()), line, col);
                }
            }
            '=' => {
                if self.match_char('=') {
                    self.emit(TokenType::EQEQ, TokenValue::Str("==".to_string()), line, col);
                } else if self.match_char('>') {
                    self.emit(TokenType::ARROW, TokenValue::Str("=>".to_string()), line, col);
                } else {
                    self.emit(TokenType::EQ, TokenValue::Str("=".to_string()), line, col);
                }
            }
            '!' => {
                if self.match_char('=') {
                    self.emit(TokenType::BANGEQ, TokenValue::Str("!=".to_string()), line, col);
                } else {
                    self.emit(TokenType::BANG, TokenValue::Str("!".to_string()), line, col);
                }
            }
            '<' => {
                if self.match_char('=') {
                    self.emit(TokenType::LTEQ, TokenValue::Str("<=".to_string()), line, col);
                } else {
                    self.emit(TokenType::LT, TokenValue::Str("<".to_string()), line, col);
                }
            }
            '>' => {
                if self.match_char('=') {
                    self.emit(TokenType::GTEQ, TokenValue::Str(">=".to_string()), line, col);
                } else {
                    self.emit(TokenType::GT, TokenValue::Str(">".to_string()), line, col);
                }
            }
            '&' => {
                if self.match_char('&') {
                    self.emit(TokenType::AMPAMP, TokenValue::Str("&&".to_string()), line, col);
                } else {
                    self.emit(TokenType::AMP, TokenValue::Str("&".to_string()), line, col);
                }
            }
            '|' => {
                if self.match_char('>') {
                    self.emit(TokenType::PIPE, TokenValue::Str("|>".to_string()), line, col);
                } else if self.match_char('|') {
                    self.emit(TokenType::BARBAR, TokenValue::Str("||".to_string()), line, col);
                } else {
                    self.emit(TokenType::BAR, TokenValue::Str("|".to_string()), line, col);
                }
            }
            '?' => {
                if self.match_char('.') {
                    self.emit(TokenType::QQDOT, TokenValue::Str("?.".to_string()), line, col);
                } else if self.match_char('?') {
                    self.emit(TokenType::QQ, TokenValue::Str("??".to_string()), line, col);
                } else {
                    self.emit(TokenType::QUESTION, TokenValue::Str("?".to_string()), line, col);
                }
            }
            '+' => {
                if self.match_char('=') {
                    self.emit(TokenType::PLUS_EQ, TokenValue::Str("+=".to_string()), line, col);
                } else {
                    self.emit(TokenType::PLUS, TokenValue::Str("+".to_string()), line, col);
                }
            }
            '-' => {
                if self.match_char('=') {
                    self.emit(TokenType::MINUS_EQ, TokenValue::Str("-=".to_string()), line, col);
                } else if self.match_char('>') {
                    self.emit(TokenType::THIN_ARROW, TokenValue::Str("->".to_string()), line, col);
                } else {
                    self.emit(TokenType::MINUS, TokenValue::Str("-".to_string()), line, col);
                }
            }
            '*' => {
                if self.match_char('*') {
                    self.emit(TokenType::STARSTAR, TokenValue::Str("**".to_string()), line, col);
                } else if self.match_char('=') {
                    self.emit(TokenType::STAR_EQ, TokenValue::Str("*=".to_string()), line, col);
                } else {
                    self.emit(TokenType::STAR, TokenValue::Str("*".to_string()), line, col);
                }
            }
            '/' => {
                if self.match_char('=') {
                    self.emit(TokenType::SLASH_EQ, TokenValue::Str("/=".to_string()), line, col);
                } else {
                    self.emit(TokenType::SLASH, TokenValue::Str("/".to_string()), line, col);
                }
            }
            '%' => self.emit(TokenType::PERCENT, TokenValue::Str("%".to_string()), line, col),
            '(' => self.emit(TokenType::LPAREN, TokenValue::Str("(".to_string()), line, col),
            ')' => self.emit(TokenType::RPAREN, TokenValue::Str(")".to_string()), line, col),
            '{' => self.emit(TokenType::LBRACE, TokenValue::Str("{".to_string()), line, col),
            '}' => self.emit(TokenType::RBRACE, TokenValue::Str("}".to_string()), line, col),
            '[' => self.emit(TokenType::LBRACKET, TokenValue::Str("[".to_string()), line, col),
            ']' => self.emit(TokenType::RBRACKET, TokenValue::Str("]".to_string()), line, col),
            ',' => self.emit(TokenType::COMMA, TokenValue::Str(",".to_string()), line, col),
            ':' => self.emit(TokenType::COLON, TokenValue::Str(":".to_string()), line, col),
            ';' => self.emit(TokenType::SEMI, TokenValue::Str(";".to_string()), line, col),
            '#' => self.emit(TokenType::HASH, TokenValue::Str("#".to_string()), line, col),
            _ => {} // skip unknown
        }
        true
    }

    fn tokenize(&mut self) -> Vec<Token> {
        while self.pos < self.source.len() {
            // Handle indentation at line start
            if self.col == 1 {
                // Skip blank lines
                if self.source[self.pos] == '\n' {
                    self.advance();
                    continue;
                }
                // Skip comment-only lines
                if self.pos + 1 < self.source.len() && self.source[self.pos] == '/' && self.source[self.pos + 1] == '/' {
                    self.skip_line_comment();
                    continue;
                }
                // Measure and emit INDENT/DEDENT tokens
                let indent = self.measure_indent();
                let current = *self.indent_stack.last().unwrap_or(&0);
                if indent > current {
                    self.indent_stack.push(indent);
                    let line = self.line;
                    let col = self.col;
                    self.emit(TokenType::INDENT, TokenValue::Num(indent as f64), line, col);
                } else if indent < current {
                    while self.indent_stack.len() > 1 && *self.indent_stack.last().unwrap_or(&0) > indent {
                        self.indent_stack.pop();
                        let line = self.line;
                        let col = self.col;
                        self.emit(TokenType::DEDENT, TokenValue::Num(indent as f64), line, col);
                    }
                }
            }
            self.tokenize_one();
        }

        // Close any remaining open indents
        while self.indent_stack.len() > 1 {
            self.indent_stack.pop();
            let line = self.line;
            let col = self.col;
            self.emit(TokenType::DEDENT, TokenValue::Num(0.0), line, col);
        }

        let line = self.line;
        let col = self.col;
        self.emit(TokenType::EOF, TokenValue::None, line, col);
        std::mem::take(&mut self.tokens)
    }
}

pub fn lex(source: &str, filename: &str) -> Vec<Token> {
    let mut lexer = Lexer::new(source, filename);
    lexer.tokenize()
}

File: src/lexer.js
'use strict'

const { T, KEYWORDS, Token } = require('./tokens')

// String escape sequences: hoisted to avoid per-character object allocation
const _STR_ESCAPES = { n: '\n', t: '\t', r: '\r', '\\': '\\', '"': '"', "'": "'", '{': '{' }

// Hoisted character-class tests for hot lexer loops
const _RE_DIGIT_UNDERSCORE = /[0-9_]/
const _RE_DIGIT = /[0-9]/
const _RE_IDENT_START = /[a-zA-Z_$]/
const _RE_IDENT_CONT = /[a-zA-Z0-9_$]/
const _RE_AT_CONT = /[a-zA-Z0-9_]/
const _RE_HEX_DIGIT = /[0-9a-fA-F_]/


class Lexer {
  constructor(source, filename = '<input>') {
    this.source = source
    this.filename = filename
    this.pos = 0
    this.line = 1
    this.col = 1
    this.tokens = []
    // Indentation stack for template mode
    this.indentStack = [0]
    this.pendingDedents = 0
  }

  error(msg) {
    throw new SyntaxError(`${this.filename}:${this.line}:${this.col}: ${msg}`)
  }

  peek(offset = 0) {
    return this.source[this.pos + offset]
  }

  advance() {
    const ch = this.source[this.pos++]
    if (ch === '\n') { this.line++; this.col = 1 }
    else this.col++
    return ch
  }

  match(expected) {
    if (this.source[this.pos] === expected) {
      this.advance()
      return true
    }
    return false
  }

  emit(type, value, line, col) {
    this.tokens.push(new Token(type, value, line ?? this.line, col ?? this.col))
  }

  // Skip whitespace but NOT newlines (newlines are significant in template mode)
  skipInlineWhitespace() {
    while (this.pos < this.source.length && (this.source[this.pos] === ' ' || this.source[this.pos] === '\t')) {
      this.advance()
    }
  }

  skipLineComment() {
    while (this.pos < this.source.length && this.source[this.pos] !== '\n') {
      this.advance()
    }
  }

  // Measure indentation of the current line
  measureIndent() {
    let indent = 0
    while (this.pos < this.source.length) {
      const ch = this.source[this.pos]
      if (ch === ' ') { indent++; this.advance() }
      else if (ch === '\t') { indent += 2; this.advance() }
      else break
    }
    return indent
  }

  tokenizeString(quote) {
    const line = this.line
    const col = this.col - 1
    const chars = []

    while (this.pos < this.source.length) {
      const ch = this.source[this.pos]

      if (ch === '\\') {
        this.advance()
        const esc = this.advance()
        chars.push(_STR_ESCAPES[esc] ?? esc)
        continue
      }

      if (quote === '"' && ch === '{') {
        // Start interpolation
        this.emit(T.STRING, chars.join(''), line, col)
        chars.length = 0
        this.advance() // consume {
        this.emit(T.INTERP_START, '{', this.line, this.col)
        // Tokenize the expression until matching }
        this.tokenizeInterpolation()
        this.emit(T.INTERP_END, '}', this.line, this.col)
        continue
      }

      if (ch === quote) {
        this.advance()
        break
      }

      if (ch === '\n' && quote !== '`') {
        this.error('Unterminated string literal')
      }

      chars.push(ch)
      this.advance()
    }

    this.emit(T.STRING, chars.join(''), line, col)
  }

  tokenizeInterpolation() {
    // Tokenize until we hit the matching } (depth tracking)
    let depth = 0
    while (this.pos < this.source.length) {
      const ch = this.source[this.pos]
      if (ch === '{') { depth++; this.tokenizeOne(); continue }
      if (ch === '}') {
        if (depth === 0) { this.advance(); return }
        depth--
        this.tokenizeOne()
        continue
      }
      this.tokenizeOne()
    }
    this.error('Unterminated string interpolation')
  }

  tokenizeNumber() {
    const start = this.pos - 1
    // Hex literal: 0x...
    if (this.source[start] === '0' && (this.source[this.pos] === 'x' || this.source[this.pos] === 'X')) {
      this.advance() // consume x
      while (this.pos < this.source.length && _RE_HEX_DIGIT.test(this.source[this.pos])) this.advance()
      const raw = this.source.slice(start, this.pos).replace(/_/g, '')
      this.emit(T.NUMBER, parseInt(raw, 16))
      this.tokens[this.tokens.length - 1].raw = raw
      return
    }
    while (this.pos < this.source.length && _RE_DIGIT_UNDERSCORE.test(this.source[this.pos])) {
      this.advance()
    }
    if (this.source[this.pos] === '.' && _RE_DIGIT.test(this.source[this.pos + 1])) {
      this.advance()
      while (this.pos < this.source.length && _RE_DIGIT_UNDERSCORE.test(this.source[this.pos])) {
        this.advance()
      }
    }
    const raw = this.source.slice(start, this.pos).replace(/_/g, '')
    this.emit(T.NUMBER, parseFloat(raw))
    this.tokens[this.tokens.length - 1].raw = raw
  }

  tokenizeIdent(first) {
    const start = this.pos - 1  // first char already consumed
    while (this.pos < this.source.length && _RE_IDENT_CONT.test(this.source[this.pos])) {
      this.advance()
    }
    const value = this.source.slice(start, this.pos)
    const kwType = KEYWORDS.get(value)
    if (kwType) {
      this.emit(kwType, value)
    } else {
      this.emit(T.IDENT, value)
    }
  }

  tokenizeAtSign() {
    // @ can start: @state, @build, @computed, @server, @worker, @get, @ident
    const start = this.pos - 1  // @ already consumed
    while (this.pos < this.source.length && _RE_AT_CONT.test(this.source[this.pos])) {
      this.advance()
    }
    this.emit(T.AT_IDENT, this.source.slice(start, this.pos))
  }

  // Tokenize one token at the current position
  tokenizeOne() {
    this.skipInlineWhitespace()
    if (this.pos >= this.source.length) {
      this.emit(T.EOF, null)
      return false
    }

    const ch = this.advance()
    const line = this.line
    const col = this.col - 1

    // Newline: significant in template mode
    if (ch === '\n') {
      this.emit(T.NEWLINE, '\n', line, col)
      return true
    }

    // Comments
    if (ch === '/' && this.peek() === '/') {
      this.skipLineComment()
      return true
    }

    // Strings
    if (ch === '"' || ch === "'") {
      this.tokenizeString(ch)
      return true
    }

    // Numbers
    if (_RE_DIGIT.test(ch)) {
      this.tokenizeNumber()
      return true
    }

    // Identifiers and keywords
    if (_RE_IDENT_START.test(ch)) {
      this.tokenizeIdent(ch)
      return true
    }

    // @ sign (annotations and instance properties)
    if (ch === '@') {
      this.tokenizeAtSign()
      return true
    }

    // Multi-char operators
    switch (ch) {
      case '.':
        if (this.peek() === '.' && this.peek(1) === '.') { this.advance(); this.advance(); this.emit(T.SPREAD, '...'); break }
        if (this.peek() === '.' && this.peek(1) === '=') { this.advance(); this.advance(); this.emit(T.DOTDOTEQ, '..='); break }
        if (this.peek() === '.') { this.advance(); this.emit(T.DOTDOT, '..'); break }
        this.emit(T.DOT, '.')
        break
      case '=':
        if (this.peek() === '=') { this.advance(); this.emit(T.EQEQ, '=='); break }
        if (this.peek() === '>') { this.advance(); this.emit(T.ARROW, '=>'); break }
        this.emit(T.EQ, '=')
        break
      case '!':
        if (this.peek() === '=') { this.advance(); this.emit(T.BANGEQ, '!='); break }
        this.emit(T.BANG, '!')
        break
      case '<':
        if (this.peek() === '<') { this.advance(); if (this.peek() === '=') { this.advance(); this.emit(T.LSHIFT_EQ, '<<='); } else { this.emit(T.LSHIFT, '<<'); } break }
        if (this.peek() === '=') { this.advance(); this.emit(T.LTEQ, '<='); break }
        this.emit(T.LT, '<')
        break
      case '>':
        if (this.peek() === '>') { this.advance(); if (this.peek() === '=') { this.advance(); this.emit(T.RSHIFT_EQ, '>>='); } else { this.emit(T.RSHIFT, '>>'); } break }
        if (this.peek() === '=') { this.advance(); this.emit(T.GTEQ, '>='); break }
        this.emit(T.GT, '>')
        break
      case '^':
        if (this.peek() === '=') { this.advance(); this.emit(T.CARET_EQ, '^='); break }
        this.emit(T.CARET, '^')
        break
      case '&':
        if (this.peek() === '&') { this.advance(); this.emit(T.AMPAMP, '&&'); break }
        this.emit(T.AMP, '&')
        break
      case '|':
        if (this.peek() === '>') { this.advance(); this.emit(T.PIPE, '|>'); break }
        if (this.peek() === '|') { this.advance(); this.emit(T.BARBAR, '||'); break }
        this.emit(T.BAR, '|')
        break
      case '?':
        if (this.peek() === '.') { this.advance(); this.emit(T.QQDOT, '?.'); break }
        if (this.peek() === '?') { this.advance(); this.emit(T.QQ, '??'); break }
        this.emit(T.QUESTION, '?')
        break
      case '+':
        if (this.peek() === '=') { this.advance(); this.emit(T.PLUS_EQ, '+='); break }
        this.emit(T.PLUS, '+')
        break
      case '-':
        if (this.peek() === '=') { this.advance(); this.emit(T.MINUS_EQ, '-='); break }
        if (this.peek() === '>') { this.advance(); this.emit(T.THIN_ARROW, '->'); break }
        this.emit(T.MINUS, '-')
        break
      case '*':
        if (this.peek() === '*') { this.advance(); this.emit(T.STARSTAR, '**'); break }
        if (this.peek() === '=') { this.advance(); this.emit(T.STAR_EQ, '*='); break }
        this.emit(T.STAR, '*')
        break
      case '/':
        if (this.peek() === '=') { this.advance(); this.emit(T.SLASH_EQ, '/='); break }
        this.emit(T.SLASH, '/')
        break
      case '%': this.emit(T.PERCENT, '%'); break
      case '(': this.emit(T.LPAREN, '('); break
      case ')': this.emit(T.RPAREN, ')'); break
      case '{': this.emit(T.LBRACE, '{'); break
      case '}': this.emit(T.RBRACE, '}'); break
      case '[': this.emit(T.LBRACKET, '['); break
      case ']': this.emit(T.RBRACKET, ']'); break
      case ',': this.emit(T.COMMA, ','); break
      case ':': this.emit(T.COLON, ':'); break
      case ';': this.emit(T.SEMI, ';'); break
      case '#': this.emit(T.HASH, '#'); break
      default:
        this.error(`Unexpected character: ${JSON.stringify(ch)}`)
    }
    return true
  }

  tokenize() {
    while (this.pos < this.source.length) {
      // Handle blank lines and indentation at line start
      if (this.col === 1) {
        // Skip blank lines
        if (this.source[this.pos] === '\n') {
          this.advance()
          continue
        }
        // Skip comment-only lines
        if (this.source[this.pos] === '/' && this.source[this.pos + 1] === '/') {
          this.skipLineComment()
          continue
        }
        // Measure and emit INDENT/DEDENT tokens
        const indent = this.measureIndent()
        const current = this.indentStack[this.indentStack.length - 1]
        if (indent > current) {
          this.indentStack.push(indent)
          this.emit(T.INDENT, indent)
        } else if (indent < current) {
          while (this.indentStack.length > 1 && this.indentStack[this.indentStack.length - 1] > indent) {
            this.indentStack.pop()
            this.emit(T.DEDENT, indent)
          }
        }
      }

      this.tokenizeOne()
    }

    // Close any remaining open indents
    while (this.indentStack.length > 1) {
      this.indentStack.pop()
      this.emit(T.DEDENT, 0)
    }

    this.emit(T.EOF, null)
    return this.tokens
  }
}

module.exports = { Lexer }
